src/v/archival/ntp_archiver_service.cc

```diff
@@ -1943,6 +1943,10 @@ ss::future<bool> ntp_archiver::do_upload_local(
         vlog(
           _rtclog.info,
           "archival metadata replicated but manifest is not re-uploaded");
+    } else {
+        // Write to archival_metadata_stm to mark our updated clean offset
+        // as a result of uploading the manifest successfully.
+        co_await maybe_flush_manifest_clean_offset();
```

Comment:
Yes, we probably should.

This is all a bit messy, because we have a mixture of the prompt update paths and some fallback for flushing in the background if we haven't done it in a while (we don't want to flush every time around the upload loop, because during saturation the flush should always be done inline during upload).

I've updated the main commit to flush from the retention path as well as the adjacent segment commit path, and also an extra commit with an apologetic commit message that adds a fallback "flush if we haven't flushed for a manifest upload interval" path that makes it easier to have confidence that we will never get stuck.

Reply:
I agree with the approach. Like your last commit message says: "it's complicated".

Reply:
I still think that we should make the STM smarter and push some of this logic out of the upload loop. The way I see it, the STM is what updates our state and it wouldn't be unnatural for it to be aware of when it performed the last update and skip in certain cases. I know Andrew doesn't agree though ;)